neuralmonkey/encoders/transformer.py

```diff
@@ -119,63 +131,113 @@ def __init__(self,
-        ff_input = tf.contrib.layers.layer_norm(
-            self_context + prev_layer.temporal_states, begin_norm_axis=2)
+    def self_attention_sublayer(
+            self, prev_layer: TransformerLayer) -> tf.Tensor:
+        # Layer normalization
+        normalized_states = layer_norm(prev_layer.temporal_states)
 
-        # Feed-forward network hidden layer + ReLU + dropout
+        # Run self-attention
+        self_context, _ = attention(
+            queries=normalized_states,
+            keys=normalized_states,
+            values=normalized_states,
+            keys_mask=prev_layer.temporal_mask,
+            num_heads=self.n_heads,
+            dropout_callback=lambda x: dropout(
+                x, self.attention_dropout_keep_prob, self.train_mode))
+
+        # Apply dropout
+        self_context = dropout(
+            self_context, self.dropout_keep_prob, self.train_mode)
+
+        # Add residual connections
+        return self_context + prev_layer.temporal_states
+
+    def feedforward_sublayer(self, layer_input: tf.Tensor) -> tf.Tensor:
+        # Layer normalization
+        normalized_input = layer_norm(layer_input)
+
+        # Feed-forward network hidden layer + ReLU
         ff_hidden = tf.layers.dense(
-            ff_input, self.ff_hidden_size, activation=tf.nn.relu,
-            name="ff_hidden_{}".format(level))
-        ff_hidden_drop = dropout(
-            ff_hidden, self.dropout_keep_prob, self.train_mode)
+            normalized_input, self.ff_hidden_size, activation=tf.nn.relu,
+            name="hidden_state")
+
+        # Apply dropout on hidden layer activations
+        ff_hidden = dropout(ff_hidden, self.dropout_keep_prob, self.train_mode)
 
-        # Feed-forward output projection + dropout
+        # Feed-forward output projection
         ff_output = tf.layers.dense(
-            ff_hidden_drop, self.model_dimension,
-            name="ff_out_{}".format(level))
+            ff_hidden, self.model_dimension, name="output")
+
+        # Apply dropout on feed-forward output projection
         ff_output = dropout(ff_output, self.dropout_keep_prob, self.train_mode)
 
-        # Residual connections + layer normalization
-        output_states = tf.contrib.layers.layer_norm(
-            ff_input + ff_output, begin_norm_axis=2)
+        # Add residual connections
+        return ff_output + layer_input
+
+    def layer(self, level: int) -> TransformerLayer:
+        # Recursive implementation. Outputs of the zeroth layer
+        # are normalized inputs.
+        if level == 0:
+            return TransformerLayer(self.encoder_inputs, self.temporal_mask)
+
+        # Compute the outputs of the previous layer
+        prev_layer = self.layer(level - 1)
+
+        with tf.variable_scope("layer_{}".format(level - 1)):
+            with tf.variable_scope("self_attention"):
+                self_context = self.self_attention_sublayer(prev_layer)
+
+            with tf.variable_scope("feedforward"):
```

Comment:
no, tady žádný `tf.AUTO_REUSE` neni, tak předpokládám, že by to šlo vyhodit i shora.